### PR TITLE
LSP: Simplify parser and parse tree for LSP

### PR DESCRIPTION
**Pull request by jmcardon**
This PR is a prelude that's necessary for further LSP work. It essentially simplifies the parser and moves a lot of the work that was previously done in the tokenizer + parser into the desugarer.

The parse tree is thus simplified significantly and far more independent of what we need for execution, which will help our LSP provide more accurate information on hover (the LSP analyzer is coming in a subsequent PR)

It also improves repl docs. That is, pre this pr:

```
pact>with-capability
(interactive):1:0: Unexpected input after expr, remaining tokens: [with-capability]
 1 | with-capability
   | ^^^^^^^^^^^^^^^
```

Post:
```
pact>with-capability
## with-capability

Use `with-capability` to apply the access to a specific capability to execute a body of code.
This function ensures that an elevated privilege—defined as a capability using a `defcap` code block—is present during the execution of the provided body of code. 

You can only call the `with-capability` function in the same module that contains the corresponding `defcap` declaration. 
If the token that grants permission to use the specified capability isn't found, the `with-capability`  evaluates the capability definition to install or grant the permission token. 
The permission token is automatically revoked after executing the code body. 
Nested `with-capability` calls for the same permission token detect the presence of the token and execute the body without reapplying the capability.

...
```

PR checklist:

* [x] Test coverage for the proposed changes
* [x] PR description contains example output from repl interaction or a snippet from unit test output
* [x] (If Relevant) Documentation has been (manually) updated at https://docs.kadena.io/pact

Additionally, please justify why you should or should not do the following:

* [x] Benchmark regressions
  - Benchmarks show no difference in execution speed
* [x] Confirm replay/back compat (Ignore until core release)
* [x] (For Kadena engineers) Run integration-tests against a Chainweb built with this version of Pact (Ignore until core release)

### CODE REVIEW

File: pact-tests/gas-goldens/builtinGas.golden [hunk 1]
```diff
@@ -81,7 +81,7 @@ not: 139
 not?: 139
 or?: 139
 pact-id: 60000271
-pairing-check: 67003096
+pairing-check: 67003097
```
Comment (rsoeldner) [inline pact-tests/gas-goldens/builtinGas.golden:84]: I'm wondering what has caused the pairing-check gas increase?

File: pact-tests/pact-tests/yield.repl [hunk 2]
```diff
@@ -92,7 +92,7 @@
 
 ;; check events
 (expect "step 0 events"
- [{"module-hash": "oKFbzkuEYAFhP-S2mW7hRvRJUPJf2FvMFy1CpxhGs4o"
+ [{"module-hash": "rmN99MpBmJbapgVRV3GjII6I4UUkl8k5pBF7-k92Lt8"
```
Comment (rsoeldner) [inline pact-tests/pact-tests/yield.repl:95]: isn't the module-hash based on stable encoding? It shouldn't change?

File: pact/Pact/Core/Syntax/LexUtils.hs [hunk 3]
```diff
@@ -80,26 +79,18 @@ data Token
...
-  | TokenEnforceOne
   | TokenSingleTick !Text
   | TokenIdent !Text
   | TokenNumber !Text
   | TokenString !Text
   | TokenTrue
   | TokenFalse
-  | TokenBlockIntro
-  | TokenSuspend
+  -- | TokenBlockIntro
```
Comment (rsoeldner) [inline pact/Pact/Core/Syntax/LexUtils.hs:88]: Cleanup

File: pact/Pact/Core/Syntax/ParseTree.hs [hunk 4]
```diff
@@ -377,32 +367,29 @@ data CapForm i
...
+  | Lam [MArg i] (NonEmpty (Expr i)) i
   | App (Expr i) [Expr i] i
-  | Block (NonEmpty (Expr i)) i
-  | Operator Operator i
   | List [Expr i] i
   | Constant Literal i
-  | Try (Expr i) (Expr i) i
-  | Suspend (Expr i) i
   | Object [(Field, Expr i)] i
+  -- | Block (NonEmpty (Expr i)) i
```
Comment (rsoeldner) [inline pact/Pact/Core/Syntax/ParseTree.hs:389]: same

**Review by rsoeldner — APPROVED** [review]

File: pact-tests/pact-tests/yield.repl (hunk 2 above)
Comment (jmcardon) [reply pact-tests/pact-tests/yield.repl:95]: It's not based on stable encoding. It _should_ change. It's based on our CBOR encoding.

**Pull request merged**